### ISSUE TITLE
Prevent select2:open when clearing selection

### ISSUE THREAD

**Issue by kartik-v** [select2/select2#3320]
Prevent catching the click (and disallow opening of the select or triggering the `select2::open`) when the clear indicator is pressed to clear the selection when `allowClear` is `true`.

This behavior worked fine in 3.5.2 - not sure if there is a reason for this. 

Use case: When `allowClear` is `true` and one clicks the selection clear indicator -- the value gets cleared but it also triggers the dropdown to open again when it is not necessary.

**Comment by sectus** [comment]
This feature "makes" weird solutions to prevent it: http://stackoverflow.com/questions/29618382/disable-dropdown-opening-on-select2-clear

**Comment by kartik-v** [comment]
I did manage to implement this by trapping the `select2:opening` and `select2:unselect` events.

**Comment by kreintjes** [comment]
I can confirm this issue.

I thinks this issue should be considered a bug, since when a user wants to clear the select, he does not immediately want to select another value. And if he does, he would simply click on the element anywhere but the clear button.

**Comment by kreintjes** [comment]
@kartik-v Could you please explain what you do exactly? I tried cancelling the opening event using e.preventDefault(). This works, but I do get a JavaScript error `TypeError: args is undefined`.

My code as I now have it:
```
  $field.on 'select2:unselecting', ->
      $(this).data('unselecting', true)
    $field.on 'select2:opening', (e) ->
      if $(this).data('unselecting')
        $(this).removeData('unselecting')
        e.preventDefault()
```

@kevin-brown What is the reason for the new behaviour? Is it possible to restore the old behaviour, potentially as an option?

**Comment by kevin-brown** [comment]
> This works, but I do get a JavaScript error TypeError: args is undefined.

Known issue, tracking at https://github.com/select2/select2/issues/3431.

> What is the reason for the new behaviour?

Based on the commit (https://github.com/select2/select2/commit/ffed37013d3c8af77c3777d71d45abc10649dbb4), there wasn't any major reason for it.

> Is it possible to restore the old behaviour, potentially as an option?

I'm open to the idea of it, though it would have to be done in a way such that it didn't break backwards compatibility.

**Comment by kartik-v** [comment]
> Could you please explain what you do exactly? I tried cancelling the opening event using e.preventDefault(). This works, but I do get a JavaScript error TypeError: args is undefined.

I had the below hack - which you can check on my [Yii2 Select2 widget demo pages](http://demos.krajee.com/widget-details/select2) - but not sure maybe I also face the args undefined message probably:

```js
// initialize a variable like this
var isClearClicked = false;
// trap these events
$field.on('select2:opening', function(e) {
    if (window['isClearClicked']) {
        event.preventDefault();
        window['isClearClicked'] = false;
    }
}).on('select2:unselect', function(e) {
     window['isClearClicked'] = true;
});
```

**Comment by kartik-v** [comment]
I modified the hack to the following code combining all the thoughts in this thread and it seems to work without errors:

```js
var $el = $('#select-id');
$el.on('select2:opening', function(e) {
    if ($el.data('unselecting')) {    
        $el.removeData('unselecting');
        setTimeout(function() {
            $el.select2('close')
        }, 10);
    }
}).on('select2:unselect', function(e) {
    $el.data('unselecting', true)
});
```

**Comment by kartik-v** [comment]
Just to add - I need to also implicitly clear the value of the element and trigger change so the modified code for the `unselecting` event is below

```js
$el.on('select2:unselecting', function(e) {
    $el.data('unselecting', true).val('').trigger('change');
})
```

**Comment by kartik-v** [comment]
Another side effect of clear - #3452 - we probably need additional events.

**Comment by NovaFocus** [comment]
The following works for me:
			$("select").on("select2:unselecting", function (e) {
				$(this).select2("val", "");
				e.preventDefault();
			});

**Comment by SlyDave** [comment]
> Is it possible to restore the old behaviour, potentially as an option?
> I'm open to the idea of it, though it would have to be done in a way such that it didn't break backwards compatibility.

I agree with others, this is odd default behaviour - it doesn't match the default behaviour of the standard multiselect HTML element on Windows (Chrome, FF and standard Office and Windows UI tested) or MAC, and in our focus group test of our new UI, it confused and frustrated users (as they had to do an additional click to access options below the select2 element after they had cleared all the selected items from the multiselect)

Further to this is is a change in behaviour from Select2 v3, so yet another incompatibility - with no evidence supported reason for the change :(

I would like to see the default behaviour changed to match v3, or at the very least a option for it.

**Comment by SectorNine50** [comment]
Here's another hack for those that can't stand the quick flick that shows up in kartik's solution (full credit to him for his solution being a springboard to this solution):
```
var select2Obj = $('select');
select2Obj.select2();

/* The following mess is a result of Select2's default behavior of triggering the
 * opening event twice upon unselecting an item. */
select2Obj.on('select2:unselecting', function(e) {
    $(this).data('unselecting1', true);
    $(this).data('unselecting2', true);
});
select2Obj.on('select2:open', function(e) {
    var unselecting1 = $(this).data('unselecting1');
    var unselecting2 = $(this).data('unselecting2');
    
    if(unselecting1 || unselecting2) {
        $(this).select2('close');
        
        if(unselecting1) {
            $(this).data('unselecting1', false);
        } else {
            $(this).data('unselecting2', false);
        }
    }
});
// End of mess.
```

**Comment by althaus** [comment]
:+1: for making it to not open the dropdown.

Maybe an option like "autoOpenOnClear" could solve the issue. Or people could trigger the opening theirselves with the unselecting event?

**Comment by Hlsgs** [comment]
@SectorNine50 your solution works, but disables the normal opening of the dropdown for one click after unselecting from the dropdown list, so i came up with this, witch i think works better because it prevents the open event from firing in the first place:
```javascript
var select2Obj = $('select');
select2Obj.select2();

select2Obj.on('select2:unselecting', function(e) {
    $(this).on('select2:opening', function(e) {
        e.preventDefault();
    });

});
select2Obj.on('select2:unselect', function(e) {
     var sel = $(this);
     setTimeout(function() {
       sel.off('select2:opening');
     }, 1);
});
```

Enjoy :)

**Comment by SectorNine50** [comment]
@Hlsgs Odd, it doesn't have that behavior on my end...  Both of the ```unselecting``` "variables" should be unset immediately after deselecting an object.  Does one remain set to true after deselecting on your end?

That said, what you came up with is a good solution as well!

**Comment by Hlsgs** [comment]
@SectorNine50 unfortunately, i don't know how to watch for variables that are not defined straight in js. btw, how would one watch for unselecting1 and unselecting2 in your code?

regardless, i think this behaviour is because unselecting by clicking the "x" in teh selection actiually reopens the dropdown but undelecting from the dropdonw does not. so after unselecting we're left with one more round of preventing teh dropdown from opening.

i've dug a little deeper here, and decided to fix select2.js as per here 6be96cfaa175448ef70b978648de7d6cb5822e89 and use preventdefault() as TypeError: args is undefined is a select2 issue and this is going to be in future releases anyway
after that, i wanted to go with @kreintjes 's solution as it seemed teh most to the point to me, but that didn't work i think because the open event gets fired twice(witch you accounted for by using two variables, correct?)
so i made a hybrid of your and his solution like this
```javasccript
$(this).on('select2:unselecting', function(e) {
    $(this).data('unselecting1', true);
    $(this).data('unselecting2', true);
});
$(this).on('select2:opening', function(e) {
    var unselecting1 = $(this).data('unselecting1');
    var unselecting2 = $(this).data('unselecting2');

    if(unselecting1 || unselecting2) {

        e.preventDefault();

        if(unselecting1) {
            $(this).data('unselecting1', false);
        } else {
            $(this).data('unselecting2', false);
        }
    }
});
```
This produced exactly the same results as your vanilla solution, and after deselecting from the dropdown the opening of it gets blocked for a round.

In the i went back to my solution, as it doesn't rely on setting variable and, by hooking into the :unselect event to restore opening functionality it disregards wether the dropdown was actually opened or not. And the timeout accouns for the opening event firing more than once(or not).

Considering all this, i should be fine with this solution even if in future updates open fires only once and/or the developer prevents the opening of the dropdown in the original code. Correct?

**Comment by SectorNine50** [comment]
>unfortunately, i don't know how to watch for variables that are not defined straight in js. btw, how would one watch for unselecting1 and unselecting2 in your code?

Within your debugger, you should be able to dig into the parent variable ```$(this)``` and find the ```data``` object, which should contain the ```unselected``` variables.

>Considering all this, i should be fine with this solution even if in future updates open fires only once and/or the developer prevents the opening of the dropdown in the original code. Correct?

If future versions only fire open once or not at all, you'll have to change the code, since one (or both) variable(s) will still remain set, meaning that the next attempt to open will be blocked.

I'm not entirely sure what your patch changes, honestly.  I'll have to dig a little more into the core code to get an understanding.

**Comment by straps** [comment]
+1 for this...great component anyway

**Comment by Geczy** [comment]
```
                $(this).select2("val", "");
                e.preventDefault();
```
Works great, thanks

**Comment by sirNemanjapro** [comment]
Still no solution?

**Comment by kevin-brown** [comment]
The planned solution is to add a `select2:clear` event that will contain the `originalEvent`, so we can just call `stopPropagation()` on that to prevent the opening. That would make the solution similar to https://github.com/select2/select2/issues/3209#issuecomment-149663474.

**Comment by kartik-v** [comment]
@SectorNine50 and others... with rel v4.0.1...  I have modified the hack slightly to the following. It now seemingly works without the flicker.

```js
var $el = $('#select-id');
$el.on('select2:unselecting', function(e) {
    $el.data('unselecting', true);
}).on('select2:open', function(e) {
    if ($el.data('unselecting')) {    
        $el.removeData('unselecting');
        $el.select2('close');
    }
});
```

**Comment by SectorNine50** [comment]
@kartik-v Very nice, that's much more concise!

**Comment by Jeromche** [comment]
@kartik-v Thanks, that works.

+1 On separating `unselect` and `open`.

**Comment by althaus** [comment]
@kartik-v Cheers for the snippet!

**Comment by nkamenar** [comment]
If anyone is still have issues with this here is a simple clear solution that is working for me:

```
$("Selector").on("select2:unselecting", function (e) {
    $(this).select2("val", "");
    e.preventDefault();
});
```

**Comment by urkle** [comment]
@nkamenar that solution has side-effects however..  With that change the "val","" call will clear out ALL selections in a multi-select if you are simply wanting to unselect a single option.

**Comment by urkle** [comment]
@kartik-v your solution seems the best..  though you don't need the "$el" var as you can use $(this).
```javascript
$('.select2').select2().on("select2:unselecting", function (e) {
    $(this).data('unselecting', true);
  }).on('select2:open', function(e) {
    if ($(this).data('unselecting')) {
      $(this).select2('close').removeData('unselecting');
    }
  });
```

**Comment by Phibbs14** [comment]
@nkamenar's solution now prevents select2 from clearing it's selected value in 4.0.3.
@urkle/@kartik-v - this solution is working for me.

**Comment by Bigdragon13th** [comment]
@nkamenar This is not working anymore on 4.0.3

**Comment by tkw722** [comment]
Hi All, for what it's worth I thought I'd add yet another snippet to this thread! It's a variant of others that have been used. This snippet avoids any blinking of the select opening relying on preventDefault to stop the open event.

```typescript
$el.on('select2:opening', (e) => {
    if ($el.data('unselecting')) {    
        $el.removeData('unselecting');
        e.preventDefault();
    }
}).on('select2:unselecting', (e) => {
    $el.data('unselecting', true);
});
```

**Comment by Tersoal** [comment]
I confirm that @urkle option works like a charm.

**Comment by ScottTrenda** [comment]
@Hlsgs: the code in your last snippet from October 28th doesn't work because you're removing the `select2:opening` event handler on the wrong event. `select2:unselect` fires immediately after `select2:unselecting`, before `select2:opening` fires. I do like your approach the best, though. Directly manipulating the event handlers seems much cleaner than setting a temp data property, in my eyes.

I took that approach, set it on the correct event handlers, and extended it to fix a similar issue for multi-selects (the dropdown closes if it's open and you remove an item). Here's the code:

```
var $element = $('select')

// Select2 4.0 incorrectly toggles the dropdown when clearing and removing options.
// Fix this by canceling the select2:opening/closing events that occur immediately after a select2:unselect event.
$element.on('select2:unselect', function() {
  function cancelAndRemove(event) {
    event.preventDefault()
    $element.off('select2:opening', cancelAndRemove)
    $element.off('select2:closing', cancelAndRemove)
  }
  $element.on('select2:opening', cancelAndRemove)
  $element.on('select2:closing', cancelAndRemove)
})
```

**Comment by Hlsgs** [comment]
@ScottTrenda Hmm, I think I'm going to use your code aswell as i dislike the closing of the multiselect dropdown while unselecting. Thanks!

So how would you correct the code i had so as to work without requiering a timeout?

**Comment by ScottTrenda** [comment]
By adding your handlers to the `select2:unselect` and `select2:opening` events instead of the `select2:unselecting` and `select2:unselect` events, respectively.

**Comment by ScottTrenda** [comment]
I updated the code above slightly, to remove the event handlers in a timeout as well as on the events. The immediate (incorrect) opening/closing events don't fire when unselect happens via keyboard backspace, and the previous code was preventing the next (correct) opening/closing event afterward in that case.

**Comment by ericmdantas** [comment]
If someone is using angular 1.x, here's how I wrapped @kartik-v's answer in a directive:

```js
;(function(ng) {
  "use strict";

  ng.module('myApp')
    .directive('select2AllowClearFix', [function() {
        return function(scope, element, attrs) {
          element.on('select2:unselecting', function(e) {
              element.data('unselecting', true);
          }).on('select2:open', function(e) {
              if (element.data('unselecting')) {
                  element.removeData('unselecting');
                  element.select2('close');
              }
          });
        }
    }]);
}(angular));
```

And then use it like:

```html
<select multiple
        class="my-select"
        select2-allow-clear-fix
        style="width: 100%"
        ng-model="model.info"
        ng-options="info as info.name for info in InfoCtrl.infos track by info.id"></select>
```

**Comment by asdsteven** [comment]
Why not just remove line 65 this.trigger('toggle', {});
in /src/js/select2/selection/allowClear.js?

**Comment by creage** [comment]
After digging into code it seems that setting a `disabled` option to `true` also prevents dropdwon from opening.

`this.inputElement.on('select2:unselect', () => {
        this.inputElement.data('select2').options.set('disabled', true);
	setTimeout(() => {
		this.inputElement.data('select2').options.set('disabled', false);
	}, 0);
});`

**Comment by ericmdantas** [comment]
@asdsteven that's not a good solution, because you should never touch the dist files.

**Comment by PersonThing** [comment]
The solution from @creage seems best so far.

Most other solutions break this scenario:
1. select an option
2. open dropdown and pick the same option again (unselects it)
3. open dropdown (prevented, because the previous step did not fire extraneous open events, so this one was prevented instead)
4. open dropdown (works again)

**Comment by elmart** [comment]
Solution from @creage breaks main scenario for me:
- Open dropdown and pick an option different than the current one.
- The dropdown doesn't close.

I'm using 4.0.3.

**Comment by kartik-v** [comment]
Yes this solution suggested works without needing to tamper with the `open` event. Phew...

```js
var opts = $(this).data('select2').options;
opts.set('disabled', true);
setTimeout(function() {
    opts.set('disabled', false);
}, 1);
```

**Comment by taai** [comment]
Hey, @kartik-v, I found a fix by @phlax : https://github.com/translate/pootle/pull/5590

But I figured out a much shorter and less error-prone solution 😉 

```javascript
$('#select-id').on('select2:unselecting', function() {
    $(this).one('select2:opening', function(ev) { ev.preventDefault(); });
});
```

I'm using Select2 version 4.0.4

**Comment by alexweissman** [comment]
It's worth mentioning that, as of 4.0.6, `unselect` and `clear` will be separate events.  See #5058.  Can someone confirm whether or not this resolves the problem (or makes it worse? 😨 )

**Comment by taai** [comment]
@alexweissman What do you mean with "whether or not this resolves the problem" – do you mean my code or the pull request you are mentioning? I just tested version 4.0.6-rc.1 – my code still works. But if you are asking whether the pull request resolves the issue we are trying to overcome, then no, in the version 4.0.6-rc.1 this issue still exists.

I can tell you how to fix the issue, if you want to:
Just remove ```this.trigger('toggle', {});``` in https://github.com/select2/select2/blob/4.0.6-rc.1/dist/js/select2.js#L1893 to prevent opening when clearing (single mode). Why is it there anyway?! 😀 
And add ```evt.stopPropagation();``` after line 1690 https://github.com/select2/select2/blob/4.0.6-rc.1/dist/js/select2.js#L1690 to prevent opening when removing (multiple mode).

**Comment by taai** [comment]
Until it gets fixed... I found out a little bit more effective way to overcome (to hack) the issue. I found out that you can access the original event when unselecting (in multiple mode), so you don't have to bind a single-use event, you can just ```stopPropagation()``` the original event. But for clearing (in single mode) you still have to bind single-use event, because it calls ```this.trigger('toggle', {});``` no matter what you do (for now).

```javascript
$('#select-id').on('select2:unselecting', function(ev) {
    if (ev.params.args.originalEvent) {
        // When unselecting (in multiple mode)
        ev.params.args.originalEvent.stopPropagation();
    } else {
        // When clearing (in single mode)
        $(this).one('select2:opening', function(ev) { ev.preventDefault(); });
    }
});
```

I think it should work for all 4.0.x versions. I have tested it for versions 4.0.4 and 4.0.5.

**Comment by alexweissman** [comment]
@taai it sounds like you are much more intimately familiar with the issue than I am.  Would you consider submitting a pull request?

**Comment by alexweissman** [comment]
Also, if you look at @kevin-brown's last comment on this:

> The planned solution is to add a `select2:clear` event that will contain the `originalEvent`, so we can just call `stopPropagation()` on that to prevent the opening. That would make the solution similar to #3209 (comment).

So, now that we have a `select2:clear` event, we should probably implement it this way.

The alternative is to introduce another boolean config option, but I really hate introducing more boolean config options.  The more we add, the more unintended interaction effects tend to crop up.

**Comment by taai** [comment]
@alexweissman I don't think this should be prevented by binding an event – for Select2 users (developers) it is an overkill just to get rid of weird behavior of Select2. If opening after clearing/unselecting is really needed, then we should implement a boolean config option... But personally I think that the behavior (opening after clearing/unselecting) is weird. By the way, in "multiple" mode this behavior was just a side effect – the click just propagated to other ```click``` bindings – and was not intended in the first place. If it should open, in code it should be executed specifically, like that ```this.trigger('toggle', {});``` which I removed in my pull request. 😉

**Comment by piotrekkmt** [comment]
Not sure if this is related, but should the clear button disable the V on the dropdowns? Is this a config issue on our side maybe or is it a flaw in select2?

![image](https://user-images.githubusercontent.com/1800333/33841487-c57938a4-de8f-11e7-9354-9cbe88ad92bb.png)

**Comment by taai** [comment]
@piotrekkmt No, it should be like that. It means you can either clear/unselect by clicking on that ```X``` or select another value by clicking on that ```V```. Otherways, to select another value, you would have to click more than once – starting with ```X```, then clicking ```V```... Yeah, probably the behavior of all this should be configurable, but the way it is right now makes sense to me. 😉

**Comment by Apezdr** [comment]
My approach!

```
$('select').on("select2:unselecting", function (e) { 
	// make sure we are on the list and not within input box
	if (e.params._type === 'unselecting') {
			$(this).val('').trigger('change');
			e.preventDefault();
	}
});
```

**Comment by ryancastle** [comment]
@Apezdr I think that ends method leaves the <option> elements selected inside the <select>, although that wouldn't matter if you weren't using them.

**Comment by seilsepor** [comment]
find this keyword in select2.js file:
if ($.inArray(name, preventableEvents) === -1) { return; }

then write this code below it :
if (name == 'unselecting') { this.$element.val(null).trigger('change'); evt.preventDefault(); }

**Comment by maruara** [comment]
select2 version 4.0.5

$('select')
.on('select2:opening', function(e) {
    if($(this).data('unselecting')) {
        $(this).removeData('unselecting');
        if(!$(this).data('isOpen')) {
            e.preventDefault();
        }
    }
})
.on('select2:unselecting', function(e) {
    $(this).data({
        unselecting: true,
        isOpen: $(this).data('select2').isOpen()
    });
});

**Comment by rnd0101** [comment]
Is there any official fix for this, for instance, some configuration option? It's utterly strange that dropdown is opened on unselecting an item (or all items) at least when allowing empty selection.

**Comment by radioactiveboyx** [comment]
@urkle your solution works tested  on  Select2 4.0.6-rc.1

**Comment by ZakAroui** [comment]
I had another issue in addition to the one described here (opening after clearing/unselecting). 
Issue is: Having two select2 dropdowns, if I open one of them and hit the clear button on the other dropdown, then the one that's open stays open, and the one where I clicked "clear" opens up as well, so I end up with both dropdowns open, and I cannot close them unless I choose an item in each one of them (sometime scrolling the page closes them). So  Based on @taai's work-around, I added another work-around to close any other opened select2 dropdowns. This is not an efficient solution though. Feedback is appreciated!

```javascript
$('select').on('select2:unselecting', function(ev) {
            if (ev.params.args.originalEvent) {
	        ev.params.args.originalEvent.stopPropagation();
	    } else {
	    	$("select").each(function( index ) {
				if ($(this).hasClass("select2-hidden-accessible") && $(this).data('select2').isOpen()) {
					$(this).select2('close');
				}
			});
	        $(this).one('select2:opening', function(ev) { ev.preventDefault(); });
	    }
	});
```

**Comment by reckart** [comment]
Apparently a solution has been suggested - is there a chance this could be integrated into the codebase?

**Comment by kartik-v** [comment]
Yup waiting for the solution to be merged.

**Comment by taai** [comment]
Hey, @kevin-brown @fk @jpic, please merge my PR #5160 which fixes this issue! I posted it in Dec 9, 2017 and since then you haven't made a decision or resolved the issue in some other way... Here is a test case: https://jsbin.com/sacunojefe/edit?js,output

**Comment by kevin-brown** [comment]
I'm pretty sure this is something that can be done now natively with the newly added `select2;clear` event. Or at least you can use that event to add a temporary event handler which will block the `select2:opening` event and get what you are looking for.

**Comment by Hlsgs** [comment]
Please don't close this if unresolved.

**Comment by gorj-tessella** [comment]
Still not fixed in 4.0.8

**Comment by kevin-brown** [comment]
Yup, this can be done natively. I followed the guidance at https://github.com/select2/select2/issues/3320#issuecomment-162606914 and made a jsbin for everyone.

https://jsbin.com/duyisusawo/edit?html,js,output

```js
$("select").on("select2:clear", function (evt) {
  $(this).on("select2:opening.cancelOpen", function (evt) {
    evt.preventDefault();
    
    $(this).off("select2:opening.cancelOpen");
  });
});
```

I'm going to turn this ticket into a documentation ticket, so this snippet can be added to the documentation for others to use.

**Comment by kartik-v** [comment]
Thanks. Is there a reason this feature cannot be included within the plugin by default instead of one needing to write this code for every select2 plugin render?

**Comment by kevin-brown** [comment]
> Is there a reason this feature cannot be included within the plugin by default instead of one needing to write this code for every select2 plugin render?

Yes, there's actually a pretty solid reason: maintenance.

The snippet takes a few minutes to copy and install globally (you don't need to copy/paste it for each one, it'll only affect `<select>` boxes with Select2 initialized). If we were to bring this in natively to Select2, we would have to add _yet another boolean option_ which would have to work with the internal event model to accomplish the same goal. Inevitably that would introduce complex, hard-to-test bugs with other supported options, which then means I have to hunt those down and add further tests for it.

The external event system exists for others to be able to integrate with Select2 and allow it to do things which are not natively supported. This is an awesome example of one of those cases.

**Comment by gorj-tessella** [comment]
Another boolean option is not needed if the default behavior were to not open the dialogue on clear, which seems much more sensible especially for multiselect. If I click clear I mean clear, not clear and open just so I have to close it. As it stands I have to double-check that every single time I setup a select2 control (often dynamically, on many pages) I also ensure this event hook is attached.

**Comment by wildlyinaccurate** [comment]
I'm surprised that this is a "feature". We switched to select2 recently and testing feedback was that this issue is a bug. Kind of disappointing that I have to wrangle events just to get around this behaviour.

**Comment by handpc** [comment]
> The following works for me:
> 
> ```js
>             $("select").on("select2:unselecting", function (e) {
>                 $(this).select2("val", "");
>                 e.preventDefault();
>             });
> ```

Simple and worked!
Thank you! You save my time :)

**Comment by nafetsrybak** [comment]
> 
> 
> Yup, this can be done natively. I followed the guidance at [#3320 (comment)](https://github.com/select2/select2/issues/3320#issuecomment-162606914) and made a jsbin for everyone.
> 
> https://jsbin.com/duyisusawo/edit?html,js,output
> 
> ```js
> $("select").on("select2:clear", function (evt) {
>   $(this).on("select2:opening.cancelOpen", function (evt) {
>     evt.preventDefault();
>     
>     $(this).off("select2:opening.cancelOpen");
>   });
> });
> ```
> 
> I'm going to turn this ticket into a documentation ticket, so this snippet can be added to the documentation for others to use.

Hello, yes it worked for me, but unfortunately it still sends AJAX request after unselecting the value(
How can i solve this problem?

**Comment by kevin-brown** [comment]
@nafetsrybak that sounds like a different issue and possibly a bug, any chance you can open a new ticket about it?

**Comment by Martin12350** [comment]
This is definitely not feature, but bug. If we will ignore fact that clear means clear and not adding as people mentioned before, here is picture what will happen if somebody clear multiple selects without @NovaFocus fix.
![bug](https://user-images.githubusercontent.com/47942899/107919725-14243380-6f6c-11eb-9f8e-ce504276deb6.png)

Unfortunatelly even with @NovaFocus fix the problem is not solved. It works only for selects that are simple, not multiple. In that case when somebody wanna remove only one option, he will delete all options except the last one.

Update: for those who wants clear single and multiple without dropdown here is my little trick (still need some trick when removing by backspace):
```
$('select').on('select2:unselecting', function (e) {
	if(e.params.args.originalEvent===undefined) {
		e.preventDefault();
		$(this).select2("val", "");
	}
});
```

**Comment by drj613** [comment]
> > Yup, this can be done natively. I followed the guidance at [#3320 (comment)](https://github.com/select2/select2/issues/3320#issuecomment-162606914) and made a jsbin for everyone.
> > [jsbin.com/duyisusawo/edit?html,js,output](https://jsbin.com/duyisusawo/edit?html,js,output)
> > ```js
> > $("select").on("select2:clear", function (evt) {
> >   $(this).on("select2:opening.cancelOpen", function (evt) {
> >     evt.preventDefault();
> >     
> >     $(this).off("select2:opening.cancelOpen");
> >   });
> > });
> > ```
> > 
> > 
> > I'm going to turn this ticket into a documentation ticket, so this snippet can be added to the documentation for others to use.

This works for me for preventing the dropdown from re-opening when you click clear, but prevents the dropdown from opening again at all after that.

---

This solution worked best for me:

https://stackoverflow.com/a/35807365

```js
$('#my-select').select2({
    allowClear: true
}).on('select2:unselecting', function() {
    $(this).data('unselecting', true);
}).on('select2:opening', function(e) {
    if ($(this).data('unselecting')) {
        $(this).removeData('unselecting');
        e.preventDefault();
    }
});
```

**Comment by reimerdes** [comment]
> Yup, this can be done natively. I followed the guidance at [#3320 (comment)](https://github.com/select2/select2/issues/3320#issuecomment-162606914) and made a jsbin for everyone.
> 
> https://jsbin.com/duyisusawo/edit?html,js,output
> 
> ```js
> $("select").on("select2:clear", function (evt) {
>   $(this).on("select2:opening.cancelOpen", function (evt) {
>     evt.preventDefault();
>     
>     $(this).off("select2:opening.cancelOpen");
>   });
> });
> ```
> 
> I'm going to turn this ticket into a documentation ticket, so this snippet can be added to the documentation for others to use.

Works for me.
If I use this on a select2 combobox the dropdown will show an directly close again. Is there a way to prevent this?

**Comment by joker-777** [comment]
I came across [this particular code snippet](https://github.com/platanus/activeadmin_addons/blob/master/app/javascript/activeadmin_addons/inputs/select2.js#L39)
```javascript
        // Related with https://github.com/select2/select2/issues/3320
        $(sallowClearelect).on('select2:unselecting', function() {
          $(this).data('unselecting', true);
        }).on('select2:open', function() {
          if ($(this).data('unselecting')) {
            $(this).select2('close').removeData('unselecting');
          }   
        });
```
which links to this ticket. Unfortunately this hack influences the behavior of the select2 tags plugin negatively. Usually, it would just start removing the last letter of the previous tag when using the backspace key. With this "hack" it removes the whole "tag" :(

**Comment by MaheryLala** [comment]
After digging for a bit I found a solution and changed a few things, idk if this'd work for y'all

```
$(".select2").on("select2:unselecting", function () {
        $(this).on("select2:opening", function (ev) {
            ev.preventDefault();
            $(this).off("select2:opening");
        });
    });
```

**Comment by pvh-archifact** [comment]
another question about select2. 

When i select select-A  it will populate select-B from ajax  but then in select-B it will default select the first item. How can i prevent it preselecting the first item from the population
